### DISCUSSION TITLE
Meets nondeterministic behavior using turmoil

### DISCUSSION THREAD

**Discussion by violin0622** [tokio-rs/turmoil#149] [Q&A] [answered]
Sorry for another question.  I've met some puzzle . I want to simulate an scenario below, all of which happens sequentially:

1. Server start
2. Client "c1" connect to server successfully.
3. Server crash
4. Another client "c2" connect to server meets timeout. 
5. Server reboot.
6. Client "c3" connect  to server successfully.

And the codes are:

```rust
#[test]
fn test_main() -> turmoil::Result {
    use rand::SeedableRng;
    use tokio::time::{timeout, Duration};
    use turmoil::net::{TcpListener, TcpStream};

    let mut sim = Builder::new()
        .epoch(SystemTime::UNIX_EPOCH)
        .rng(rand::rngs::StdRng::seed_from_u64(10))
        .ip_version(turmoil::IpVersion::V6)
        .build();

    sim.host("s1", || async {
        let sock = TcpListener::bind("[::]:80").await?;
        println!("s1 start");
        loop {
            sock.accept().await?;
        }
    });
    sim.client("c1", async {
        TcpStream::connect("s1:80").await?;
        println!("c1 connect succees");
        Ok(())
    });

    sim.step()?;
    sim.step()?;

    sim.crash("s1");
    sim.client("c2", async {
        assert!(
            timeout(Duration::from_millis(1), TcpStream::connect("s1:80"))
                .await
                .is_err()
        );
        println!("c2 connect should timeout");
        Ok(())
    });

    sim.step()?;
    sim.step()?;

    sim.bounce("s1");
    sim.client("c3", async {
        TcpStream::connect("s1:80").await?;
        println!("c1 connect succees");
        Ok(())
    });

    sim.run()
}
```

There are two questions:
- Why do I have to `sim.step()` **TWICE** rather than once to make test pass ?
- Given the deterministic epoch and rand seed, why does the code still fails occasionally ?

**Comment by mcches** [comment, marked as answer]
> Why do I have to sim.step() TWICE rather than once to make test pass ?

If you check out the documentation for [`step()`](https://github.com/tokio-rs/turmoil/blob/main/src/sim.rs#L318) you'll see that this simply steps things forward once for each host. Use `run()` to step until clients have completed.

> Given the deterministic epoch and rand seed, why does the code still fails occasionally ?

Can you switch to `run()` and we can debug further if necessary?

**Reply by violin0622** [reply]
Yes, you're right. With `sim.step()` replaced by `sim.run()` , all tests passed and the result was stabilized.
